### ISSUE TITLE
ember-wormhole error

### ISSUE THREAD

**Issue by transitorytammy** [sir-dunxalot/ember-tooltips#309]
Hi hi. 

I am getting this error randomly in production from `ember-wormhole` which is a dependency of `ember-tooltip`.  

`The head node of a wormhole must be attached to the DOM`

https://github.com/yapplabs/ember-wormhole/blob/master/addon/components/ember-wormhole.js#L63

Wondering if you had suggestions on the best way to solve this on my side?

**Comment by maxfierke** [comment]
Hi @transitorytammy! Which version(s) of ember-wormhole is resolved & installed in your app? Could you also share a sample of how `ember-tooltips` is being used when you get this error and the full stack trace, if possible?

**Comment by timmorey** [comment]
I've been running into this one too, but haven't managed to repro outside of my app.  It appears to be a timing issue, where an `ember-tooltip` gets rendered and then removed before its wormhole is fully initialized.  In particular, I'm seeing https://github.com/yapplabs/ember-wormhole/blob/master/addon/components/ember-wormhole.js#L74 run immediately before https://github.com/yapplabs/ember-wormhole/blob/master/addon/components/ember-wormhole.js#L60-L65.

So, there is some questionable behavior in my app, which is causing content to be rendered and then immediately replaced, though it would still be nice if we could avoid the crash.

I'm running 
ember-tooltips 3.0.1
ember-wormhole 0.5.5
ember-source 3.0.0 (working on upgrading to a more recent version)

More complete error (stack trace truncated, since it is incredibly long)

```
Uncaught Error: The head node of a wormhole must be attached to the DOM
    at run.schedule (ember-wormhole.js:71)
    at invoke (backburner.js:205)
    at Queue.flush (backburner.js:125)
    at DeferredActionQueues.flush (backburner.js:278)
    at Backburner.end (backburner.js:410)
    at Backburner._boundAutorunEnd (backburner.js:372)
```

**Comment by maxfierke** [comment]
@transitorytammy @timmorey Not sure if this is relevant or not: https://github.com/yapplabs/ember-wormhole/issues/71. It makes it sound like it could be nebulously related to an old `ember-cli-htmlbars` version or other dependencies, so if either of you are able to share with me your `package.json` and related lock file, that would help a lot!

Possibly a quick experiment: could you give it a shot upgrading (or try `3.0.2`, which bumped `ember-cli-htmlbars` requirement to `^3.0.0`)?

**Comment by benclum11** [comment]
I'm also experiencing the issue and it's preventing me from upgrading from 2.x to 3.0.1 or 3.0.2

**Comment by maxfierke** [comment]
Could you please share a excerpt of how you're invoking `{{ember-tooltip}}` and details about the surrounding context & when you're seeing the error? Is it when used in a component or route template? On an initial render, update, component teardown, or when entering or exiting a route?

**Comment by panthony** [comment]
@maxfierke I have the same problem and it was NOT fun to pin point.

- `ember@3.0`
- `ember-cli-htmlbars@2.0.3`
- `ember-tooltips@3.0.2`

Minimal reproduction case I could find:

```hbs
{{#each itemsList as |item|}}
    <div>
      label
      {{#ember-tooltip}}
        tooltip
      {{/ember-tooltip}}
    </div>
{{/each}}
```

If itemsList has:

- less than 11 elements -> works
- 11 elements -> Uncaught Error: infinite rendering invalidation detected
- 12 elements -> Assertion Failed: Cannot deregister unknown unregistered renderer
- more than 12 elements -> The head node of a wormhole must be attached to the DOM

**Comment by maxfierke** [comment]
Thanks @panthony! I'll take a look at this tonight!

**Comment by panthony** [comment]
@maxfierke Not related to the foreach btw.

You can't have more than 10 tooltip on a given page anywhere on the page.

**Comment by maxfierke** [comment]
@panthony hmm... that's concerning.

We do have a test case that uses more than 11 tooltips: https://github.com/sir-dunxalot/ember-tooltips/blob/master/tests/acceptance/many-tooltips-test.js#L19 and anecdotally I have pages with many more than 11 tooltips, so I'd be curious about any additional context you could provide that might point to different options or configuration that might be causing additional runloops to be scheduled.

Are you able to get a stack-trace of when it's hitting the infinite rendering invalidation?

**Comment by panthony** [comment]
@maxfierke The stacktrace is, like often, unhelpful.

![capture d ecran 2018-11-15 a 17 37 50](https://user-images.githubusercontent.com/230462/48567389-34806600-e8fd-11e8-9c66-fd1b913da908.png)

Seems like something goes wrong with paper-tabs @ ember-paper (maybe other addons too).

In an empty template, this crash:

```hbs
{{#paper-tabs borderBottom=true selected="tab" onChange=(route-action "selectTab") as |tabs|}}
  {{#tabs.tab value="tab"}}
    "hello"
  {{/tabs.tab}}
{{/paper-tabs}}

{{#each model.items as |item|}}
  <div>
    {{item.id}}
    {{#ember-tooltip}}
      aaa
    {{/ember-tooltip}}
  </div>
{{/each}}

```

This works:

```hbs
{{#each model.items as |item|}}
  <div>
    {{item.id}}
    {{#ember-tooltip}}
      aaa
    {{/ember-tooltip}}
  </div>
{{/each}}

```

This works too:

```hbs
{{#paper-tabs borderBottom=true selected="tab" onChange=(route-action "selectTab") as |tabs|}}
  {{#tabs.tab value="tab"}}
    "hello"
  {{/tabs.tab}}
{{/paper-tabs}}
```

Sorry I have to leave for the night, won't be able to provide more hints for today

**Comment by maxfierke** [comment]
@panthony I've looked into this and at least for the cases you've provided, it seems related to the use with ember-paper, as I could not reproduce it with `ember-tooltips` alone. Could you open another issue with what you've provided here to track the issue with the ember-paper interaction?

Curious: @transitorytammy @timmorey are either of you using any similar UI toolkit add-ons? e.g. ember-paper, ember-bootstrap, etc.? Or any other add-ons that depend on ember-wormhole?

**Comment by maxfierke** [comment]
@timmorey @mnoble01 @transitorytammy mind trying out #319 in your applications to see if removing ember-wormhole fixes things?

quick copy-paste for your package.json:

```javascript
"ember-tooltips": "https://github.com/sir-dunxalot/ember-tooltips.git#mf-maybe-in-element",
```

**Comment by mnoble01** [comment]
@maxfierke That branch did not work for me. 

I dug into this a little more and am getting the following error from `createTooltip`'s catch block.
```
createTooltip() {
  return new RSVP.Promise((resolve, reject) => {
    try {
      run(() => {
        ...
      });
    } catch(error) {
      console.log('ERROR', error);
      reject(error);
    }
  });
}
```

<img width="379" alt="screen shot 2018-11-19 at 12 34 32 pm" src="https://user-images.githubusercontent.com/463770/48724570-90960380-ebf7-11e8-8b21-f2ebdfacbda0.png">

which ultimately throws:
<img width="879" alt="screen shot 2018-11-19 at 1 10 14 pm" src="https://user-images.githubusercontent.com/463770/48726156-7e6a9400-ebfc-11e8-88f1-2ab8fe4ea31e.png">


Removing the outer `run` fixed the above errors and the tooltips show correctly. Not sure if this is the "fix" we're all looking for ¯\_(ツ)_/¯.

**Comment by maxfierke** [comment]
@mnoble01 oh, yeah, that makes sense. I don't think that outer `run` is at all necessary. Nor is that promise, it seems like. Could you try #320?

**Comment by mnoble01** [comment]
@maxfierke Works for me 🎉

**Comment by joelluijmes** [comment]
I also ran in this issue, when trying to add a tooltip inside a table where there are more than 10 rows. Just like @mnoble01 says, the `maybe-in-element` branch does not solve the issue. It actually breaks the tooltip in my case.

Using the latest commit (thus with merged #320) it works 👍

**Comment by timmorey** [comment]
@maxfierke I just got a chance to try out some of the recent changes.

When I run the branch for #319, I'm still able to repro a crash, though with a different error (`Cannot read property 'dispose' of null` @ `ember-tooltip-base.js:230`).

When I run the code on master, which includes #320, the crash appears to be fixed 🎉.

I then took the code for #319 and merged in master, and the crash was fixed there too.  So, it seems like #320 fixes the core problem here.  #319 may still be a nice direction to go, but its not necessary to fix this issue.

**Comment by maxfierke** [comment]
@timmorey Great to hear! We'll probably tag a new release with those changes this week. I ran into a failing test on an app late last week that I'm trying to determine if it's related to #320. I'm going to try and take a crack at getting to the bottom of that before releasing.

**Comment by maxfierke** [comment]
Just released 3.1.0, which includes #320 and sounds like it should fix this issue. I'm going to close this out, but please comment if it crops up in 3.1.0 as well.
